### PR TITLE
Contact selection

### PR DESCRIPTION
**Pull request by samisuteria**
cleans up contact selection and makes it a little simpler 

we should try to fetch all the friends at once instead of async (its very jarring waiting on them to load)

### CODE REVIEW

File: WalkingChallenge/Team.swift [hunk 1]
```diff
@@ -20,12 +20,12 @@ class Team: TableDataSource {
...
-      onMain {
-        completion(true)
-      }
-    }
+//    Facebook.getTaggableFriends(limit: .count(Team.limit)) { (friend) in
+//      self.cells.append(TeamMemberCellInfo(for: friend))
+//      onMain {
+//        completion(true)
+//      }
+//    }
```
Comment (compnerd) [inline WalkingChallenge/Team.swift]: Please just kill the dead code.

File: WalkingChallenge/ContactPicker.swift [hunk 2]
```diff
@@ -277,10 +221,15 @@ extension ContactPicker: UITableViewDelegate {
 
   func tableView(_ tableView: UITableView,
                  willSelectRowAt indexPath: IndexPath) -> IndexPath? {
-    let selected = tableView.indexPathsForSelectedRows
-
-    if selected?.count == Team.limit {
-      alert(message: "You are limited to \(Team.limit) members")
+    guard let selected = tableView.indexPathsForSelectedRows else { return nil }
```
Comment (compnerd) [inline WalkingChallenge/ContactPicker.swift]: This breaks the selection.

File: WalkingChallenge/ContactPickerCell.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  override func configure(cellInfo: CellInfo) {
+    guard let info = cellInfo as? ContactCellInfo else { return }
+    
+    name.text = info.name
+    onBackground {
+      guard
+        let url = URL(string: info.picture),
+        let data = try? Data(contentsOf: url)
+      else { return }
+      self.picture.image = UIImage(data: data)
```
Comment (compnerd) [inline WalkingChallenge/ContactPickerCell.swift]: This should be on the main thread.  I think you should rebase these changes.

File: WalkingChallenge/ContactPickerDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import Contacts
+
+class ContactDataSource: TableDataSource {
+  var cells = [CellInfo]()
+  var filter: String? { didSet { configureCells() }}
+  private var friends = [Friend]()
+  private var selectedFriends = Set<String>()
+  
+  func reload(completion: @escaping GenericBlock) {
+    fetchFriends(completion: completion)
```
Comment (compnerd) [inline WalkingChallenge/ContactPickerDataSource.swift:13]: Any reason for the extra function?

File: WalkingChallenge/ContactPickerDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  
+  func changeSelection(at indexPath: IndexPath, select: Bool) {
+    if let fbid = (cells[safe: indexPath.row] as? ContactCellInfo)?.fbid {
+      if select {
+        selectedFriends.insert(fbid)
+      } else {
+        selectedFriends.remove(fbid)
+      }
+      configureCells()
+    }
```
Comment (compnerd) [inline WalkingChallenge/ContactPickerDataSource.swift:24]: *sigh* I wish there was a more succinct way to write this.

File: WalkingChallenge/ContactPickerDataSource.swift [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    
+    var sortedFilteredFriends = friends
+    
+    switch sortOrder {
+    case .givenName:
+      sortedFilteredFriends.sort { $0.first_name < $1.first_name }
+    case .familyName:
+      sortedFilteredFriends.sort { $0.last_name < $1.last_name }
+    default: break
+    }
```
Comment (compnerd) [inline WalkingChallenge/ContactPickerDataSource.swift:55]: Nice!

File: WalkingChallenge/ContactPickerDataSource.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    
+    if let filter = filter {
+      sortedFilteredFriends = sortedFilteredFriends
+        .filter {$0.first_name.contains(filter) || $0.last_name.contains(filter)}
```
Comment (compnerd) [inline WalkingChallenge/ContactPickerDataSource.swift]: Why not `$0.display_name.contains(filter)`?  Do we not care about case insensitivity in the search?

**Review by compnerd — CHANGES_REQUESTED** [review]
Please rebase the changes, there were some important changes made at the same time.

File: WalkingChallenge/ContactPickerDataSource.swift (hunk 4 above)
Comment (samisuteria) [reply WalkingChallenge/ContactPickerDataSource.swift:13]: just for readability and if we want to do something else in reload later

**Pull request merged**